velox/exec/HashProbe.cpp

```diff
@@ -189,58 +225,276 @@ void HashProbe::initializeFilter(
+    spillInputReader_ = partition->createReader();
+    spillPartitionSet_.erase(iter);
   }
 
-  auto hashBuildResult =
-      operatorCtx_->task()
-          ->getHashJoinBridge(
-              operatorCtx_->driverCtx()->splitGroupId, planNodeId())
-          ->tableOrFuture(future);
+  VELOX_CHECK_NULL(spiller_);
+  spillInputPartitionIds_ = spillPartitionIds;
+  if (!spillInputPartitionIds_.empty()) {
+    const auto& spillConfig = spillConfig_.value();
+    spiller_ = std::make_unique<Spiller>(
+        Spiller::Type::kHashJoinProbe,
+        probeType_,
+        HashBitRange(
+            spillInputPartitionIds_.begin()->partitionBitOffset(),
+            spillInputPartitionIds_.begin()->partitionBitOffset() +
+                spillConfig.hashBitRange.numBits()),
+        spillConfig.filePath,
+        operatorCtx_->task()
+                ->queryCtx()
+                ->pool()
+                ->getMemoryUsageTracker()
+                ->maxTotalBytes() *
+            spillConfig.fileSizeFactor,
+        Spiller::spillPool(),
+        spillConfig.executor);
+    // Set the spill partitions to the corresponding ones at the build side. The
+    // hash probe operator itself won't trigger any spilling.
+    spiller_->setPartitionsSpilled(toPartitionNumSet(spillInputPartitionIds_));
+
+    spillHashFunction_ = std::make_unique<HashPartitionFunction>(
+        spiller_->hashBits(), probeType_, keyChannels_);
+    spillInputIndicesBuffers_.resize(spillHashFunction_->numPartitions());
+    rawSpillInputIndicesBuffers_.resize(spillHashFunction_->numPartitions());
+    numSpillInputs_.resize(spillHashFunction_->numPartitions(), 0);
+  }
+}
+
+void HashProbe::waitForTableBuild() {
+  checkRunning();
+  VELOX_CHECK_NULL(table_);
+
+  auto hashBuildResult = joinBridge_->tableOrFuture(&future_);
   if (!hashBuildResult.has_value()) {
-    VELOX_CHECK_NOT_NULL(future);
-    return BlockingReason::kWaitForJoinBuild;
+    VELOX_CHECK(future_.valid());
+    setState(State::kWaitForBuild);
+    return;
   }
 
   if (hashBuildResult->antiJoinHasNullKeys) {
-    // Anti join with null keys on the build side always returns nothing.
+    // Null-aware anti join with null keys on the build side without a filter
+    // always returns nothing.
     VELOX_CHECK(isNullAwareAntiJoin(joinType_));
-    finished_ = true;
-  } else {
-    table_ = hashBuildResult->table;
-    if (table_->numDistinct() == 0) {
-      // Build side is empty. Inner, right and semi joins return nothing in this
-      // case, hence, we can terminate the pipeline early.
-      if (isInnerJoin(joinType_) || isLeftSemiJoin(joinType_) ||
-          isRightJoin(joinType_) || isRightSemiJoin(joinType_)) {
-        finished_ = true;
+    // 'antiJoinHasNullKeys' flag shall only be set on the first built 'table_'
+    // and we don't expect any spilled probe data at that point.
+    VELOX_CHECK(spillPartitionSet_.empty());
+    spillPartitionSet_.clear();
```

Comment:
This statement seems redundant as we just checked that spillPartitionSet_ is empty.